src/java/org/apache/cassandra/utils/SystemInfo.java

```diff
@@ -0,0 +1,211 @@
+    private oshi.SystemInfo si ;
+
+    public SystemInfo() {
+        si = new oshi.SystemInfo();
+    }
+
+    /**
+     *
+     * @return The name of the current platform. (e.g. Linux)
+     */
+    public String platform()
+    {
+        return oshi.SystemInfo.getCurrentPlatform().name();
+    }
+
+    /**
+     * Gets the maximum number of processes the user can create.
+     * Note: if not on a Linux system this always return the
+     * @{code DEFAULT_MAX_PROCESSES} value.
+     * @return The maximum number of processes.
+     * @see #DEFAULT_MAX_PROCESSES
+     */
+    long getMaxProcess()
+    {
+        // this check only works on Linux systems.
+        if (oshi.SystemInfo.getCurrentPlatform() == PlatformEnum.LINUX)
+        {
+            String path = format("/proc/%s/limits", getPid());
+            try
+            {
+                List<String> lines = FileUtils.readLines(new File(path));
+                for (String line : lines)
+                {
+                    if (line.startsWith("Max processes"))
+                    {
+                        String[] parts = SPACES_PATTERN.split(line);
+
+                        if (parts.length < 3)
+                            continue;
+
+                        String limit = parts[2];
+                        return "unlimited".equals(limit) ? INFINITY : Long.parseLong(limit);
+                    }
+                }
+                logger.error("'Max processes' not found in " + path);
+            }
+            catch (Throwable t)
+            {
+                logger.error(format("Unable to read %s", path), t);
+            }
+        }
+
+        /* return the default value for non-Linux systems.
+         * can not return 0 as we know there is at least 1 process (this one) and
+         * -1 historically represents infinity.
+         */
+        return DEFAULT_MAX_PROCESSES;
+    }
+
+    /**
+     *
+     * @return The maximum number of open files allowd to the current process/user.
+     */
+    long getMaxOpenFiles()
+    {
+        // ulimit -H -n
+        return si.getOperatingSystem().getCurrentProcess().getHardOpenFileLimit();
+    }
+
+    /**
+     * Gets the Virtual Memory Size (VSZ). Includes all memory that the process can access, including memory that is swapped out and memory that is from shared libraries.
+     * @return The amount of virtual memory allowed to be allocatedby the current process/user.
+     */
+    long getVirtualMemoryMax() {
+        return si.getOperatingSystem().getCurrentProcess().getVirtualSize();
+    }
+
+    /**
+     *
+     * @return The amount of swap space allocated on the system.
+     */
+    long getSwapSize() {
+        return si.getHardware().getMemory().getVirtualMemory().getSwapTotal();
+    }
+
+    /**
+     * @return the PID of the current system.
+     */
+    public long getPid() {
+        return si.getOperatingSystem().getProcessId();
+    }
+
+    /**
+     * Tests if the system is running in degraded mode.
```

Comment:
This javadoc is not accurate anymore, it will return possible degradations as one sentence and it is not logging them.